### ISSUE TITLE
Logging gradients and a possible bug in the docs?

### ISSUE THREAD

**Issue by rbracco** [PyTorchLightning/pytorch-lightning#2660]
TL;DR Question:  How do we correctly log gradients to tensorboard? 

I am trying to log gradients to tensorboard to track NaN loss in a speech application. The docs suggest using `on_after_backward` along with this code that appears to be incorrect:

```
# example to inspect gradient information in tensorboard
if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
    params = self.state_dict()
    for k, v in params.items():
        grads = v
        name = k
        self.logger.experiment.add_histogram(tag=name, values=grads,
                                                             global_step=self.trainer.global_step)
```

This code seems to log the weights instead of gradients (assuming lightning state_dict is the same structure as pytorch). I'm happy to fix it and submit a PR as long as I'm not mistaken. I would log the weights like this...

```
# example to inspect gradient information in tensorboard
if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
    params = self.state_dict()
    for name, weight in params.items():
         self.logger.experiment.add_histogram(tag=name, values=weight,
                                                             global_step=self.trainer.global_step)
```

but for gradients, if I try to log weights.grad, I get `None`. How do we correctly log gradients to tensorboard? 
```
# example to inspect gradient information in tensorboard
if self.trainer.global_step % 25 == 0:  # don't make the tf file huge
     params = self.state_dict()
     for name, weight in params.items():
           self.logger.experiment.add_histogram(tag=name, values=weight.grad,
                                                             global_step=self.trainer.global_step)
```

**Comment by miccio-dk** [comment]
for posterity, here's an article explaining how to do it: https://www.learnopencv.com/tensorboard-with-pytorch-lightning/

**Comment by rbracco** [comment]
@miccio-dk, nice article but it doesn't talk about logging gradients. 

I figured this out and the docs are definitely incorrect, I'm happy to PR if there's interest. Anyways on to the good stuff, how to log gradients in lightning:

Do it manually:
```
def on_after_backward(self):
    global_step = self.global_step
    self.logger.experiment.add_histogram("cnn", self.model.cnn.weight, global_step)
    self.logger.experiment.add_histogram("cnn grad", self.model.cnn.weight.grad, global_step)
    self.logger.experiment.add_histogram("fully connected", self.model.fully_connected.weight, global_step)
    self.logger.experiment.add_histogram("fully connected grad", self.model.fully_connected.weight.grad, global_step)
    self.logger.experiment.add_histogram("classifier", self.model.classifier[0].weight, global_step)
    self.logger.experiment.add_histogram("classifier grad", self.model.classifier[0].weight.grad, global_step)
```
 
- cnn, fully connected...etc are the names defined in my model, you'll need to change it for yours or just log every layer with

```
def on_after_backward(self):
    global_step = self.global_step
    for name, param in self.model.named_parameters():
        self.logger.experiment.add_histogram(name, param, global_step)
        if param.requires_grad:
            self.logger.experiment.add_histogram(f"{name}_grad", param.grad, global_step)
```

**Comment by ydcjeff** [comment]
@rbracco sorry for late reply, are you still interested in sending PR?
